pkg/scheduler/cache/cache.go

```diff
@@ -1284,29 +1204,73 @@ func (sc *SchedulerCache) processBindTask() {
+	executedPreBinders := make([]PreBinder, 0, len(sc.preBinderRegistry.preBinders))
+
+	sc.preBinderRegistry.mu.RLock()
+	defer sc.preBinderRegistry.mu.RUnlock()
+
+	for _, preBinder := range sc.preBinderRegistry.preBinders {
+		if preBinder == nil {
+			continue
+		}
+
+		if err := preBinder.PreBind(ctx, bindContext); err != nil {
+			// If PreBind fails, rollback the executed PreBinders
+			for i := len(executedPreBinders) - 1; i >= 0; i-- {
+				if executedPreBinders[i] != nil {
+					executedPreBinders[i].PreBindRollBack(ctx, bindContext)
+				}
+			}
+			return err
+		}
+		executedPreBinders = append(executedPreBinders, preBinder)
+	}
+
+	return nil
+}
+
+// executePreBinds executes PreBind for a list of bindContexts
+func (sc *SchedulerCache) executePreBinds(ctx context.Context, bindContexts []*BindContext) []*BindContext {
+	logger := klog.FromContext(ctx)
+	successfulBindContexts := make([]*BindContext, 0, len(bindContexts))
+
+	for _, bindContext := range bindContexts {
+		if err := sc.executePreBind(ctx, bindContext); err != nil {
+			reason := fmt.Sprintf("execute prebind failed: %v", err)
+			klog.Error(reason)
+			if updateErr := sc.taskUnschedulable(bindContext.TaskInfo, schedulingapi.PodReasonSchedulerError, reason, ""); updateErr != nil {
+				logger.Error(updateErr, "Failed to update pod status", "pod", klog.KObj(bindContext.TaskInfo.Pod))
+			}
+			continue
+		}
+		successfulBindContexts = append(successfulBindContexts, bindContext)
+	}
+
+	return successfulBindContexts
+}
+
 // BindTask do k8s binding with a goroutine
 func (sc *SchedulerCache) BindTask() {
-	klog.V(5).Infof("batch bind task count %d", len(sc.bindCache))
-	var tmpBindCache []*schedulingapi.TaskInfo = make([]*schedulingapi.TaskInfo, len(sc.bindCache))
+	klog.V(5).Infof("batch bind task count %d", sc.batchNum)
+	tmpBindCache := make([]*BindContext, len(sc.bindCache))
 	copy(tmpBindCache, sc.bindCache)
-	go func(tasks []*schedulingapi.TaskInfo) {
-		successfulTasks := make([]*schedulingapi.TaskInfo, 0)
-		for _, task := range tasks {
-			if err := sc.VolumeBinder.BindVolumes(task, task.PodVolumes); err != nil {
-				klog.Errorf("task %s/%s bind Volumes failed: %#v", task.Namespace, task.Name, err)
-				sc.VolumeBinder.RevertVolumes(task, task.PodVolumes)
-				sc.resyncTask(task)
-			} else {
-				successfulTasks = append(successfulTasks, task)
-				klog.V(5).Infof("task %s/%s bind Volumes done", task.Namespace, task.Name)
-			}
-		}
 
-		bindTasks := make([]*schedulingapi.TaskInfo, len(successfulTasks))
-		copy(bindTasks, successfulTasks)
-		sc.Bind(bindTasks)
+	// Currently, bindContexts only contain 1 element.
+	go func(bindContexts []*BindContext) {
+		logger := klog.Background()
+		ctx := klog.NewContext(context.Background(), logger)
+		cancelCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		successfulPreBindContexts := sc.executePreBinds(cancelCtx, bindContexts)
```

Comment:
When `sc.executePreBinds` failed, `sc.Bind` will not be executed, and also `resyncTask` will not be executed, is this reasonable?

Reply:
OK I have added resyncTask inside the executePreBinds if PreBind of the task failed